src/main/java/slimeknights/tconstruct/library/modifiers/modules/technical/MaxArmorLevelModule.java

```diff
@@ -0,0 +1,102 @@
+import slimeknights.tconstruct.library.modifiers.modules.ModifierModule;
+import slimeknights.tconstruct.library.module.HookProvider;
+import slimeknights.tconstruct.library.module.ModuleHook;
+import slimeknights.tconstruct.library.tools.capability.TinkerDataCapability;
+import slimeknights.tconstruct.library.tools.capability.TinkerDataCapability.TinkerDataKey;
+import slimeknights.tconstruct.library.tools.context.EquipmentChangeContext;
+import slimeknights.tconstruct.library.tools.nbt.IToolStackView;
+import slimeknights.tconstruct.library.utils.IdParser;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+/**
+ * Module for keeping track of the max level of a modifier across all pieces of equipment.
+ * @see TinkerDataKey
+ * @see slimeknights.tconstruct.library.modifiers.modules.behavior.ShowOffhandModule
+ */
+public record MaxArmorLevelModule(TinkerDataCapability.ComputableDataKey<ModifierMaxLevel> key, boolean allowBroken, @Nullable TagKey<Item> heldTag) implements HookProvider, EquipmentChangeModifierHook, ModifierModule {
+  private static final List<ModuleHook<?>> DEFAULT_HOOKS = HookProvider.<MaxArmorLevelModule>defaultHooks(ModifierHooks.EQUIPMENT_CHANGE);
+  public static final RecordLoadable<MaxArmorLevelModule> LOADER = RecordLoadable.create(
+    new IdParser<>(ResourceLocation::new, "Stat").flatXmap(location -> TinkerDataCapability.ComputableDataKey.of(location, ModifierMaxLevel::new), TinkerDataKey::getId).requiredField("key", MaxArmorLevelModule::key),
+    BooleanLoadable.INSTANCE.defaultField("allow_broken", false, MaxArmorLevelModule::allowBroken),
+    Loadables.ITEM_TAG.nullableField("held_tag", MaxArmorLevelModule::heldTag),
+    MaxArmorLevelModule::new);
+
+  @Override
+  public RecordLoadable<MaxArmorLevelModule> getLoader() {
+    return LOADER;
+  }
+
+  @Override
+  public List<ModuleHook<?>> getDefaultHooks() {
+    return DEFAULT_HOOKS;
+  }
+
+  @Override
+  public void onEquip(IToolStackView tool, ModifierEntry modifier, EquipmentChangeContext context) {
+    updateMaxLevelIfArmor(tool, context, key, modifier, allowBroken, heldTag);
+  }
+
+  @Override
+  public void onUnequip(IToolStackView tool, ModifierEntry modifier, EquipmentChangeContext context) {
+    updateMaxLevelIfArmor(tool, context, key, modifier, allowBroken, heldTag);
+  }
+
+
+  /* Helpers */
+
+  /**
+   * Calculate the current max value of the given maxLevel in entity modifier data for an armor modifier
+   *
+   * @param context Equipment change context
+   * @param key     Key to modify
+   * @param entry   Entry for the given armor modifier
+   * @return The current max level for key
+   */
+  public static float updateMaxLevel(EquipmentChangeContext context, TinkerDataCapability.ComputableDataKey<ModifierMaxLevel> key, ModifierEntry entry) {
+    var holder = context.getTinkerData().resolve();
+
+    if(holder.isPresent()) {
+      var data = holder.get();
+      var maxLevel = data.computeIfAbsent(key);
```

Comment:
These could just be one line, `ModifierMaxLevel maxLevel = holder.get().computeIfAbsent(key)`. Means you will feel less inclination to use `var` as you don't have to list the type of the holder.